framework/src/base/Assembly.C

```diff
@@ -1354,10 +1395,32 @@ Assembly::cacheJacobianBlock(DenseMatrix<Number> & jac_block, std::vector<dof_id
         _cached_jacobian_cols.push_back(dj[j]);
       }
   }
-
   jac_block.zero();
 }
 
+void
+Assembly::cacheJacobianBlockNonlocal(DenseMatrix<Number> & jac_block, std::vector<dof_id_type> & idof_indices, std::vector<dof_id_type> & jdof_indices, Real scaling_factor)
+{
+  if ((idof_indices.size() > 0) && (jdof_indices.size() > 0) && jac_block.n() && jac_block.m())
+  {
+    std::vector<dof_id_type> di(idof_indices);
+    std::vector<dof_id_type> dj(jdof_indices);
+    _dof_map.constrain_element_matrix(jac_block, di, dj, false);
+
+    if (scaling_factor != 1.0)
+      jac_block *= scaling_factor;
+
+    for (unsigned int i=0; i<di.size(); i++)
+      for (unsigned int j=0; j<dj.size(); j++)
+        if (jac_block(i, j) != 0.0) // no storage allocated for unimplemented jacobian terms, maintaining maximum sparsity possible
```

Comment:
This isn't really right.  Try to set/return a flag from the object saying that there were/weren't jacobian contributions from that element.